rfcs/react-components/styles-handbook.md

```diff
@@ -0,0 +1,771 @@
+
+```js
+import { makeStyles, shorthands } from '@griffel/react';
+
+const useClasses = makeStyles({
+  root: {
+    /* styles */
+  },
+  foo: {
+    /* styles */
+  },
+  bar: {
+    /* styles */
+  },
+});
+
+function Component(props) {
+  const classes = useClasses();
+  const className = mergeClasses(
+    classes.root,
+    props.foo && classes.foo /* styles specific for "foo" */,
+    props.bar && classes.bar /* styles specific for "bar" */,
+  );
+
+  /* --- */
+}
+```
+
+### Do not concatenate classes
+
+It is not possible to simply concatenate classes returned by `useClasses()` hooks. Always use `mergeClasses()` to merge classes as results of concatenation can contain duplicated classes and lead to non-deterministic results.
+
+```js
+import { makeStyles, mergeClasses } from '@griffel/react';
+
+const useClasses = makeStyles({
+  rootA: { display: 'flex' },
+  rootB: { display: 'grid' },
+});
+
+function App(props) {
+  const classes = useClasses();
+
+  // ✅ Returns "class-display-grid"
+  const correctClasses = mergeClasses(classes.rootA, classes.rootB);
+  // 🔴 Never concatenate class strings, returns "class-display-flex class-display-grid"
+  const wrongClasses = classes.rootA + ' ' + classes.rootB;
+}
+```
+
+## `makeResetStyles`
+
+Is used to define base styles in components avoid the "CSS rules explosion" problem as generates a monolithic class:
+
+```js
+import { makeResetStyles } from '@griffel/react';
+
+const useBaseClassname = makeResetStyles({
+  padding: '4px',
+  margin: '4px',
+});
+```
+
+> ⚠️ **Note:** Only one class generated by `makeResetStyles()` can be applied to an element. Otherwise, behavior will be non-deterministic as classes merging will not be done for this case and results depend on order of insertion.
+
+### Hybrid approach
+
+Our recommendation is use [`makeResetStyles` API][griffel-make-reset-styles] to define base styles for a component and keep using `makeStyles` for overrides and permutations:
+
+```jsx
+import { makeStyles, makeResetStyles, mergeClasses, shorthands } from '@griffel/react';
+
+const useBaseClassname = makeResetStyles({
+  ':hover': {
+    ...shorthands.padding('4px'),
+    /* other styles */
+  },
+  '::before': {
+    display: 'block',
+    content: "' '",
+  },
+  '@media (forced-colors: active)': {
+    ...shorthands.padding('4px'),
+    /* other styles */
+  },
+});
+// ⬇️⬇️⬇️
+// produces 1 class
+
+const useClasses = makeStyles({
+  circular: {
+    ...shorthands.borderRadius('10px'),
+  },
+  primary: {
+    color: 'pink',
+  },
+});
+// ⬇️⬇️⬇️
+// produces 4/1 classes to be conditionally applied
+
+function Component(props) {
+  const baseClassName = useBaseClassname();
+  const classes = useClasses();
+
+  const className = mergeClasses(baseClassName, props.circular && classes.circular, props.primary && classes.primary);
+
+  return <div className={className} />;
+}
+```
+
+# Advanced
```

Comment:
Is there really a distinction between basic and advanced? and it's kinda weird to have this order:

* basic
* API
* advanced

I think we should combine basic and advance to be something like 'core concepts'

Reply:
I would remove Advanced. What's in the chapter is important for the engineers to understand.

Reply:
@ling1726 @miroslavstastny 

There is a problem between there:
- we need to explain what is Atomic CSS, so a reader will know why we are counting classes
- we need to introduce APIs
- and then we can explain selectors

I was thinking that to do with that and tried to do following (a6b2c35), WDYT?

- [Introduction](#introduction)
- [Core concepts](#core-concepts)
  - [Atomic CSS](#atomic-css)
  - [`makeStyles`](#makestyles)
    - [Limitations](#limitations)
    - [Performance caveat](#performance-caveat)
  - [`mergeClasses()`](#mergeclasses)
    - [⚠️ Only combine classes with `mergeClasses`](#%EF%B8%8F-only-combine-classes-with-mergeclasses)
  - [`makeResetStyles`](#makeresetstyles)
    - [Hybrid approach (Using `makeStyles` and `makeResetStyles` together)](#hybrid-approach-using-makestyles-and-makeresetstyles-together)
  - [Understanding selector complexity](#understanding-selector-complexity)
- [Best practices](#best-practices)

Reply:
lgtm!